### PR TITLE
OSD-7842 - Only route alerts to PagerDuty that originate from a managed namespace

### PR DESCRIPTION
**Pull request by tnierman**
Completes https://issues.redhat.com/browse/OSD-7842

These changes utilize the newly generated [managed-namespaces configMap](https://github.com/openshift/managed-cluster-config/blob/master/deploy/osd-managed-resources/managed-namespaces.ConfigMap.yaml) to only route alerts from OSD namespaces to PagerDuty.

### Alerts
- Alert fires from namespace present in `managed-namespaces` CM -> Alert appears in PD
- Alert fires from namespace not present in `managed-namespaces` CM -> Alert does not appear in PD 
- Alert is firing in a managed namespace, and namespace is removed from `managed-namespaces` CM -> Alert remains in PagerDuty
- Alert is firing in unmanaged namespace, and namespace is added to `managed-namespaces` CM -> Alert appears in PD

### Namespaces
- `managed-namespaces` ConfigMap is invalid, empty, or removed -> Alerts are routed to PagerDuty from all `^openshift-.*` namespaces

### CODE REVIEW

File: pkg/controller/secret/secret_controller.go
Comment (tnierman) [inline pkg/controller/secret/secret_controller.go:570]: `yaml.Unmarshal` doesn't throw an error if the `managedNamespacesConfig` struct doesn't match, checking if the namespace list is empty seems to be the easiest way to determine if something is off with the yaml

Comment (tnierman) [reply pkg/controller/secret/secret_controller.go:571]: Since this should be noted for the future, I added a comment clarifying this.

File: pkg/controller/secret/secret_controller.go
Comment (dustman9000) [inline pkg/controller/secret/secret_controller.go:635]: Can you add some unit tests around these functions in secret_controller_test.go?

Comment (tnierman) [reply pkg/controller/secret/secret_controller.go:635]: [Unit tests added](https://github.com/openshift/configure-alertmanager-operator/pull/186/files#diff-3f3b1c34e4d392411c2522d11db21d7d89b9603de69a11f63686c664558bb375R387-R429)

Comment (dustman9000) [reply pkg/controller/secret/secret_controller.go:635]: Thanks!

File: pkg/controller/secret/secret_controller.go
Comment (dustman9000) [inline pkg/controller/secret/secret_controller.go:571]: Just a minor observation that we could DRY this code up a bit and have `managedNamespaceList` be set to the default (`redhat-*`, `kube*`, and `openshift-*`) and only check for err == nil here. Appending the managed namespaces would be the exception and you wouldn't need all of these else blocks.

**Comment by dustman9000** [comment]
Nice work @tnierman! Just some minor comments/questions, but nothing major to hold this from merging. Thanks!
/lgtm

File: pkg/controller/secret/secret_controller.go
Comment (tnierman) [reply pkg/controller/secret/secret_controller.go:544]: Looks like there's still one object we might want to alert on in `^kube.*` namespaces:
```
$ oc get namespaces | grep -e ^kube.*
kube-node-lease                                    Active   29h
kube-public                                        Active   29h
kube-system                                        Active   29h
```

```
$ oc get all -n kube-system
NAME              TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                        AGE
service/kubelet   ClusterIP   None         <none>        10250/TCP,10255/TCP,4194/TCP   29h
```

```
$ oc get all -n kube-public
No resources found in kube-public namespace.
```

```
$ oc get all -n kube-node-lease
No resources found in kube-node-lease namespace.
```

By default, I don't think OSD clusters have any `^redhat-.*` namespaces:
```
$ oc get namespaces | grep -e ^redhat-.*
$
```

Given this, I'd be in favor of not alerting on `^redhat-.*` namespaces, but I think it's desirable for us to continue to monitor `^kube.*` namespaces.

File: pkg/controller/secret/secret_controller.go
Comment (fahlmant) [inline pkg/controller/secret/secret_controller.go:482]: Minor nit, but with the watch of confimaps, 480 should also be updated

File: pkg/controller/secret/secret_controller.go
Comment (fahlmant) [inline pkg/controller/secret/secret_controller.go:488]: Now that there are multiple types being handled here, a type check, at least for the once case, should be added.

File: pkg/controller/secret/secret_controller.go
Comment (fahlmant) [inline pkg/controller/secret/secret_controller.go:526]: Does this work if the type of object that triggered the reconcile is a configmap?

File: pkg/controller/secret/secret_controller.go
Comment (tnierman) [reply pkg/controller/secret/secret_controller.go:494]: Looks like `request` is a `reconcile.Request` object, which only contains a name & namespace, so I don't know if a type check is feasible. 

The good news is that even if there is both a secret and configmap named after `cmNameManagedNamespaces`, the reconcile loop should still ignore the secret and only interact with the configmap, unless I'm misunderstanding the intention here.

File: pkg/controller/secret/secret_controller.go
Comment (tnierman) [reply pkg/controller/secret/secret_controller.go:532]: Yeah, this function will return true if, any time the reconcile loop is run, there exists a configmap named after the variable `cmNameManagedNamespaces`. 

The `reconcile` loop now runs every time a configmap or secret changes, before it only would run if a secret changes.

File: pkg/controller/secret/secret_controller.go
Comment (tnierman) [reply pkg/controller/secret/secret_controller.go:488]: Changed `"Reconciling Secret"` -> `"Reconciling Object"`

File: pkg/controller/secret/secret_controller.go
Comment (tnierman) [reply pkg/controller/secret/secret_controller.go:494]: Ignore might be the wrong word, but I don't think it would have any ill-effect. 

In a case where a configmap name matches a CAMO secret name and triggers a reconcile (or vice-versa), the request name is only used to lookup the correct objects and act on them. In this case, the reconcile _is_ called unnecessary, but it shouldn't result in a different `alertmanager-main` secret, since the same objects are used to generate it each time.

File: pkg/controller/secret/secret_controller.go
Comment (tnierman) [reply pkg/controller/secret/secret_controller.go]: Ah, I just realized you meant line 526, not 522, my bad. 
The block just continues if it can't find a secret with the request name, so it doesn't affect reconciles with the `managed-namespaces` ConfigMap. I've added a corresponding block to check for ConfigMap errors as well.

File: pkg/controller/secret/secret_controller.go
Comment (pbergene) [reply pkg/controller/secret/secret_controller.go:545]: ^redhat-.* contain managed applications.  I'm not sure if there is any reliance on routing alerts via the cluster prometheus, but worth checking with @hlipsig @apahim

Comment (hlipsig) [reply pkg/controller/secret/secret_controller.go:545]: +1 redhat-* for add-ons.

Comment (dustman9000) [reply pkg/controller/secret/secret_controller.go:545]: @cblecker - Can you confirm that we need to alert on add-ons? My understanding was that we do not want to alert on layered products/add-ons under the platform alertmanager. The thinking behind this change was to have an explicit list, not a regex.

**Comment by cblecker** [comment]
The configmap (https://github.com/openshift/managed-cluster-config/blob/master/deploy/osd-managed-resources/managed-namespaces.ConfigMap.yaml) needs to be more exhaustive before we can merge this. We need to have all the OCP namespaces in there.

We also need to ensure the following alerts will still fire/route after this:
https://github.com/openshift/ops-sop/blob/master/v4/alerts/RHMIUpgradeExpectedDurationExceeded.md
https://github.com/openshift/ops-sop/blob/master/v4/alerts/RHOAMOperatorInstallDelayed.md
https://github.com/openshift/ops-sop/blob/master/v4/alerts/RHOAMUpgradeExpectedDurationExceeded.md

File: pkg/controller/secret/secret_controller.go
Comment (fahlmant) [inline pkg/controller/secret/secret_controller.go:495]: Can this be removed?

**Review by fahlmant — APPROVED** [review]
/lgtm

**Pull request merged**
